libsol/instruction.c

```diff
@@ -18,6 +19,8 @@ enum ProgramId instruction_program_id(const Instruction* instruction, const Mess
         return ProgramIdVote;
     } else if (memcmp(program_id, &spl_token_program_id, PUBKEY_SIZE) == 0) {
         return ProgramIdSplToken;
+    } else if(memcmp(program_id, &spl_token2022_program_id, PUBKEY_SIZE) == 0) {
+        return ProgramIdSplToken;//Treat the Token2022 exactly the same as the SplToken
```

Comment:
Is the decoding resulting from this used in the user signing loop?
If so this may cause issues, as the user may be unaware that he is signing transactions using the token22 program that may contain transfer fee or transfer hooks with unintended consequences

Reply:
The main reason behind this is the code thats prints transaction details. If I detect token2022 as something different then it gets more messy to handle basic token instruction like mint, burn etc. At this point, if we use for example `TransferFeeExtension`  we just ignore it - we don't display any additional screens. 

What we can do is to throw error(instruction not supported) when we get any extension instruction. This way there will be no chance to sign something that you are not able to verify visually on the screen. Otherwise we would need to redesign this function https://github.com/blockydevs/app-solana/blob/develop/libsol/transaction_printers.c#L535 `https://github.com/blockydevs/app-solana/blob/develop/libsol/transaction_printers.c#L535` because it is not able to handle anything with more than 3 instructions and we dont know what extensions someone will send and in what order

Reply:
So, like main idea in this PR is to allow transactions that use the token2022 program id to be able to perform basic instructions that are also compatible with the old token

Reply:
So, like main idea in this PR is to allow transactions that use the token2022 program id to be able to perform basic instructions that are also compatible with the old token

Reply:
There is a slight mix up between extension setup and the extension usage.

The `TransferFeeExtension` instruction is only called to setup the extension, so not supporting it here makes total sense. However, if its setup, any calls to `Transfer` are going to need extra accounts to correctly execute the transfer with the fee. 

The way the decoding is currently setup, any Token22 mint with extensions such as Transfer Fees or Transfer Hooks enabled, are going to execute completely blind to the decoding operation due to the `Transfer` instruction being ABI compatible between both programs (extra accounts needed by extensions are appended at the end, everything else maintains the same binary structure).

At the very least, during decoding, the user should be aware that its signing a `Transfer` for a mint with such extensions enabled (fees can go all the way up to 100% and can change over time, on the other end of the spectrum transfer hooks can enable bad actors to effectively wipe accounts).

Reply:
Yes, you are right about it. So the only option I see right now is to throw unsupported instruction error when someone tries to sign transaction with the T2022 extension. This way we will avoid a situation when someone signs a transfer with some hidden fees coming form extesion that we didnt show on the screen. And at the same time we will be able to handle rest of the "basic" operations on that program without any problem

What do you think about it? @tiago18c @sponomarev (@piotrswierzy )

Reply:
I think they can be supported with little effort but with an higher focus on transparency.

Even on a perfect case scenario with RPC access to retrieve current extensions, one would still need to simulate transactions and in some scenarios (e.g. transfer hook program upgrade between simulation and tx submission) the simulations can be wrong.

In the case of transfer fees, it would be helpful to be able to display to the user the expected fee on a transfer, however I can see how this would require a major effort to support given that we can't decode this from the instruction inputs.

My suggestion is to take a lazy approach here that focus on transparency. Flag Token22 `Transfer` instructions as such rather than Token program, and advise the user to check an explorer as these may contain transfer hooks or transfer fees that the app/device can't infer the outputs strictly from the instruction inputs.

Reply:
@tiago18c Hello, so we thought how we can achieve some more transparency for the user and we added genereal warning screen if transaction contains any token2022 extension.
Below you can see how it looks on the nanox emulator for token transfer with fees extensions:
![image](https://github.com/blockydevs/app-solana/assets/142581992/bb7c6292-f870-45d5-823b-10c78ec2bc3c)

Reply:
This simple approach seems reasonable at that stage of token2022 support.

Reply:
Its much better.

However, if I understood this correctly, the warning message only pops up if there are Extensions related instructions on the TX right? For that scenario its the right behaviour.

However, if there is only a `TransferChecked` instruction in the transaction and it is of a token mint with possible damaging extensions enabled (high transfer fees, draining transfer hook), the user will think he's signing a token transfer from the old Token Program without any side effects as none of that is present in the transaction.

IMO there needs to be another warning on any T22 `TransferChecked` like:
"Token Extensions" "Cannot verify mint extensions" 
"Verify mint issuer before signing"  
(not sure of character limits, but something along these lines)

Reply:
Yes, any extension thats coming from T22 generates this warning, the idea behind this PR was to allow users to use the old backward compatible features. 
Due to the current code design we are a little bit limited in terms of detecting any complex cases with different instructions. Transaction printers contains hardcoded switch case that does not allow a lot of flexibility unfortunately.

Can you specify what exact cases should trigger that additional warning? Is it for transactions that only contains single `TransferChecked` instruction (https://solana-labs.github.io/solana-program-library/token/js/functions/createTransferCheckedInstruction.html) or are they any other cases?

And there is also another issue. On the older nanos devices we are really close to the memory limit, so adding long strings in to the binary may cause problems too, some of them appear only in specific features so maybe we just reuse the same warning text?

Reply:
The additional warning should be triggered on `TransferChecked` when `programId == token22`, regardless of how many instructions are present in the transaction.

I'd say you can recycle the title part of the warnings.

Title: "Extension warning"
Messages: "Can't check mint extensions" -> "Check explorer" 

Is the memory available enough for those? If not we can rework these warnings to further reuse strings.

Reply:
@tiago18c ok, I've added additional warning screens and i've also merged latest develop changes to verify that everything is still working as planned. 
I did few tests on the nanos model and it looks that we are still within memory limits